telemetry/tracks/class-tracks-client.php

```diff
@@ -0,0 +1,258 @@
+<?php
+/**
+ * Telemetry: Tracks Pixel class
+ *
+ * @package Automattic\VIP\Telemetry
+ */
+
+declare(strict_types=1);
+
+namespace Automattic\VIP\Telemetry;
+
+use WP_Error;
+use function Automattic\VIP\Logstash\log2logstash;
+
+/**
+ * Handles all operations related to the Tracks pixel.
+ */
+class Tracks_Client {
+	/**
+	 * Base URL of the Tracks pixel.
+	 */
+	protected const PIXEL_BASE_URL = 'https://pixel.wp.com/t.gif';
+
+	/**
+	 * Tracks REST API endpoint for post requests
+	 */
+	protected const TRACKS_ENDPOINT = 'https://public-api.wordpress.com/rest/v1.1/tracks/record';
+
+	/**
+	 * Class singleton instance.
+	 *
+	 * @var ?Tracks_Client
+	 */
+	protected static $instance = null;
+
+	/**
+	 * Events queued to be sent to the Tracks pixel.
+	 *
+	 * @var array<Tracks_Event>
+	 */
+	protected $events = array();
+
+	/**
+	 * Constructor.
+	 */
+	private function __construct() {
+		// Track events asynchronously (inject pixels in the footer).
+		add_action( 'admin_footer', array( $this, 'render_tracking_pixels' ) );
+		// Synchronously track any remaining events that could not be added to
+		// the footer.
+		add_action( 'shutdown', array( $this, 'record_remaining_events' ) );
+	}
+
+	/**
+	 * Instantiates the singleton.
+	 *
+	 * @return Tracks_Client
+	 */
+	public static function instance(): Tracks_Client {
+		if ( is_null( static::$instance ) ) {
+			static::$instance = new Tracks_Client();
+		}
+
+		return static::$instance;
+	}
+
+	/**
+	 * Enqueues an event to be recorded asynchronously.
+	 *
+	 * @param Tracks_Event $event The event to record.
+	 * @return bool|WP_Error True if the event was enqueued for recording.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if the event is generating an error.
+	 */
+	public function record_event_asynchronously( Tracks_Event $event ) {
+		$is_event_recordable = $event->is_recordable();
+
+		if ( true !== $is_event_recordable ) {
+			return $is_event_recordable;
+		}
+
+		static::instance()->events[] = $event;
+
+		return true;
+	}
+
+	/**
+	 * Records an event synchronously (using a GET request).
+	 *
+	 * @param Tracks_Event $event The event to record.
+	 * @return bool|WP_Error True if recording the event succeeded.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if any error occurred.
+	 */
+	public function record_event_synchronously( Tracks_Event $event ) {
+		$is_event_recordable = $event->is_recordable();
+
+		if ( true !== $is_event_recordable ) {
+			return $is_event_recordable;
+		}
+
+		$pixel_url = static::instance()->generate_pixel_url( $event );
+
+		if ( null === $pixel_url ) {
+			log2logstash( [
+				'severity' => 'error',
```

Comment:
Should this be an error, or warning? Considering the status code is 400 wondering if that's better? Just don't want the customer's logs filled up with errors because the plugin didn't implement an event correctly.

Reply:
If the library can't generate the proper pixel URL then the site won't be able to send events for AJAX requests. I think that's a pretty serious problem that should be logged IMHO